Add flash attention and conv2d direct controls for image generation

wbruna opened the pull request:
This is a straight implementation that adds controls and command-line switches for the flash_attention, diffusion_conv_direct and vae_conv_direct flags.

For Conv2D Direct, see https://github.com/leejet/stable-diffusion.cpp/pull/744 for details. But in a nutshell: VAE decoding can be ~2.4-3.8x faster, with less than a half memory usage (allowing 1024x1024 generations without VAE tiling). The diffusion process currently doesn't seem to benefit from it, so the main reason to enable the control would be exposing the functionality to more testing.

The main drawback is a possible crash for unsupported backends. We could in principle only honor the flag for known to be working configurations, or blacklist broken ones. On my tests, it's currently working on CPU and Vulkan, while ROCm crashes.

The separate flash attention control is useful to avoid certain bugs, like https://github.com/leejet/stable-diffusion.cpp/issues/756 , and configurations where flash attention benefits text generation, but slows down image generation.

Depends on #1669 .

wbruna committed:
Add separate flash attention config for image generation

wbruna committed:
Add config option for Conv2D Direct

LostRuins commented:
In what scenarios does it break (is it model dependent) or is it purely backend dependent? Also, are there any drawbacks to it quality/perf wise?

If it's an all round improvement, rather than overcomplicate with too many flags, would it be better to just force-enable it for supported backends? or is it conditionally broken?

wbruna commented:
It directly depends on the implementation of a specific operation, so it's unlikely to crash on a specific model (except in the sense that a model family which doesn't use that operation would of course always work, but I think all currently do).

In terms of performance, it's a bit too early to be sure. When it helps, it helps a lot, but it's not clear if it could get worse for some backend+model+hardware combinations. Quality seems to be the same, although outputs may vary a bit; in my tests, the VAE output changes less than it does with tiling.

Another possibility could be enabling the flags unconditionally for known-to-be-working cases, but including an environment variable to work around possible issues (either forcing it on to test a new backend, or off to avoid a special case).

wbruna commented:
>  it's not clear if it could get worse for some backend+model+hardware combinations

Already found an example - my own card! VAE timings for rendering a 512x512 SDXL image, radv versus amdvlk:

| | radv | amdvlk |
| :---: | :---: | :---: |
| base | 3.54s | 3.74s |
| vae-conv-direct | 0.86s | 11.46s |

It's even more dramatic for diffusion-conv-direct : ~20% worse on radv, 5x worse on amdvlk.

LostRuins commented:
Alright how about this:
- we can add the extra new toggle for SD flash attention
- we add a new toggle for conv direct, but we combine both of them into a single flag for simplicity. The flag toggles them both. How about `--sdconvdirect` (avoid 'dir' as that usually means directory here)

wbruna commented:
> we add a new toggle for conv direct, but we combine both of them into a single flag for simplicity. The flag toggles them both. How about --sdconvdirect (avoid 'dir' as that usually means directory here)

A single boolean option would be a bit annoying, to be honest :-)  My card gets 20% slower for the diffusion step, so it could be slower overall for high step counts, even if the VAE is much faster.

But a non-boolean sdconvdirect would be fine. I've implemented one with three values: disabled, enabled, and 'vaeonly' to enable it only for the VAE (I don't think we'll find a case where only the the VAE would be slower, but it's easy to add an extra value for that if you wish).

LostRuins commented:
Anyway let me know when its ready to review

wbruna commented:
Should be good to go now... except maybe for the combobox control. I tried to follow the other labeled controls as examples; it _works_, but you'll probably want to adjust its position and padding. It's being called only at one point right now, but I may make use of it in another (unrelated) PR.

LostRuins commented:
You added `sd_diffusion_convdir_var` but it is not used anywhere

wbruna commented:
Cleaned up that var, and rebased on top of `concedo_experimental`.

Edit: while working on #1692 , I started to think `on` and `off` could be better than `enabled` and `disabled`; or maybe `off`, `vaeonly` and `full`. I don't really have a strong preference, but maybe it'd be best to choose values that could be the same for other options, for consistency?

LostRuins commented:
I think the disambiguation is good. So long as the commands are clear it's good. Generally I try to avoid enabled/on/off/disabled flags because that is implied with store true, i.e. `--usemmap` means mmap is enabled. So if 3 options are needed I think keeping it explicit is ok

LostRuins reviewed (approved):
should be good, merging
